lib/src/array.h

```diff
@@ -66,6 +66,47 @@ extern "C" {
 #define array_assign(self, other) \
   array__assign((VoidArray *)(self), (const VoidArray *)(other), array__elem_size(self))
 
+#define array__search_sorted(self, start, compare, suffix, needle, index, exists) \
+  do { \
+    *(index) = start; \
+    *(exists) = false; \
+    uint32_t size = (self)->size - *(index); \
+    if (size == 0) break; \
+    int comparison; \
+    while (size > 1) { \
+      uint32_t half_size = size / 2; \
+      uint32_t mid_index = *(index) + half_size; \
+      comparison = compare(&((self)->contents[mid_index] suffix), (needle)); \
+      if (comparison <= 0) *(index) = mid_index; \
+      size -= half_size; \
+    } \
+    comparison = compare(&((self)->contents[*(index)] suffix), (needle)); \
+    if (comparison == 0) *(exists) = true; \
+    else if (comparison < 0) *(index) += 1; \
+  } while (0)
+
+#define _compare_int(a, b) ((int)*(a) - (int)(b))
```

Comment:
Can you add a comment to clarify that only `a` is a pointer? Looking at the definition of `array__search_sorted` I think that's right, but it's definitely not typical for a binary search callback.